### PR TITLE
Initial setup of codebase and creates Openshift template  

### PR DESCRIPTION
**Pull request by GeetikaBatra**
Signed-off-by: Geetika Batra gbatra@redhat.com

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
+FROM registry.centos.org / centos / centos: 7
```
Comment (msrb) [inline Dockerfile]: This looks fishy (?) What's up with the whitespaces - does this work? :)

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+WORKDIR ${APP_DIR}
+
+RUN yum install - y epel - release & &\
+    yum install - y gcc git python34 - pip python34 - devel & &\
+    yum clean all & &\
+    mkdir - p ${APP_DIR}
+
+
+COPY f8a_notification / ${APP_DIR} / f8a_notification
```
Comment (msrb) [inline Dockerfile]: Again, some weird whitespaces here. Misconfigured IDE maybe?

**Comment by GeetikaBatra** [comment]
@msrb. Please take a look

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,27 @@
+REGISTRY?= registry.devshift.net
+REPOSITORY?= fabric8 - analytics / f8a - analytics - notification
```
Comment (msrb) [inline Makefile]: Ehm O:-) Those pesky whitespaces again :wink:

File: cico_setup.sh [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    else
+    echo "Could not login, missing credentials for the registry"
+    exit 1
+    fi
+}
+
+prep() {
+    yum - y update
+    yum - y install epel - release
+    yum - y install git which python34 - pip python34 - virtualenv docker
```
Comment (msrb) [inline cico_setup.sh]: Whitespaces ;)

File: cico_setup.sh [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    docker push "${target}"
+}
+
+push_image() {
+    local image_name
+    local image_repository
+    local short_commit
+    image_name =$(make get - image - name)
+    image_repository =$(make get - image - repository)
+    short_commit =$(git rev - parse - -short=7 HEAD)
```
Comment (msrb) [inline cico_setup.sh]: Whitespaces ;)

File: detect-common-errors.sh [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+pass = 0
+fail = 0
+
+function prepare_venv() {
+    VIRTUALENV =$(which virtualenv)
+    if [$? -eq 1]
+    then
+    # python34 which is in CentOS does not have virtualenv binary
+    VIRTUALENV =$(which virtualenv - 3)
```
Comment (msrb) [inline detect-common-errors.sh]: Whitespaces ;)

File: detect-dead-code.sh [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+function prepare_venv() {
+    VIRTUALENV =$(which virtualenv)
+    if [$? -eq 1]
+    then
+    # python34 which is in CentOS does not have virtualenv binary
+    VIRTUALENV =$(which virtualenv - 3)
+    fi
+
+    ${VIRTUALENV} - p python3 venv & & source venv / bin / activate & & python3 "$(which pip3)" install vulture
```
Comment (msrb) [inline detect-dead-code.sh]: Whitespaces ;)

File: setup.cfg [hunk 8]
```diff
@@ -0,0 +1,3 @@
+[pycodestyle]
+ignore = E121, E126, E402
+max - line - length = 100
```
Comment (msrb) [inline setup.cfg]: Whitespaces ;)

File: .gitignore [hunk 9]
```diff
@@ -0,0 +1,6 @@
+notify/
```
Comment (msrb) [inline .gitignore]: Note you have some `.pyc` files in this PR. Please remove the `.pyc` files and update `.gitignore` so it won't happen again.

File: Makefile [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+DEFAULT_TAG = latest
+
+ifeq($(TARGET), rhel)
+    DOCKERFILE: = Dockerfile.rhel
+else
+    DOCKERFILE: = Dockerfile
+endif
+
+.PHONY: all docker - build fast - docker - build test get - image - name get - image - repository
```
Comment (msrb) [inline Makefile]: Same here.

File: Makefile [hunk 11]
```diff
@@ -0,0 +1,27 @@
+REGISTRY?= registry.devshift.net
+REPOSITORY?= fabric8 - analytics / f8a - analytics - notification
+DEFAULT_TAG = latest
+
+ifeq($(TARGET), rhel)
+    DOCKERFILE: = Dockerfile.rhel
```
Comment (msrb) [inline Makefile]: The RHEL Dockerfile is nowhere to be found (?)

File: Makefile [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+else
+    DOCKERFILE: = Dockerfile
+endif
+
+.PHONY: all docker - build fast - docker - build test get - image - name get - image - repository
+
+all: fast - docker - build
```
Comment (msrb) [inline Makefile]: And there is more :)

File: check-bashscripts.sh [hunk 13]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+shellcheck - e 2181 - e 1091 . / *.sh
```
Comment (msrb) [inline check-bashscripts.sh]: Whitespaces ;)

File: openshift/f8a-notification.yaml [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+    description: f8a-notification
+objects:
+- apiVersion:  batch/v2alpha1
+  kind: CronJob
+  metadata:
+    name: f8a-notification-scheduler
+    annotations:
+      description: f8a-notification-scheduler
+  spec:
+    schedule: "*/1 * * * *"
```
Comment (msrb) [inline openshift/f8a-notification.yaml]: You want to run the job every minute (?) Is that correct?

File: openshift/f8a-notification.yaml [hunk 15]
```diff
@@ -0,0 +1,44 @@
+apiVersion: v1     
+kind: Template
+labels:
+  template: f8a-notification
+metadata:
+  name: f8a-notification
+  annotations:
+    description: f8a-notification
+objects:
+- apiVersion:  batch/v2alpha1
```
Comment (msrb) [inline openshift/f8a-notification.yaml]: Would newer API `batch/v1beta1` work here?

File: openshift/helpers.sh [hunk 16]
```diff
@@ -0,0 +1,47 @@
+
+function is_set_or_fail() {
```
Comment (msrb) [inline openshift/helpers.sh]: What's the purpose of this file here?

File: f8a_notification/entryscript.py [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Author: Geetika Batra <gbatra@redhat.com>
+#
+
+"""Class to create cron script."""
+
+
+def run():
```
Comment (msrb) [inline f8a_notification/entryscript.py]: Missing docstring :D

File: openshift/f8a-notification.yaml [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+  kind: CronJob
+  metadata:
+    name: f8a-notification-scheduler
```
Comment (msrb) [inline openshift/f8a-notification.yaml]: Let's be consistent here. `f8a-notification` or `f8a-notification-scheduler`? :)

File: openshift/f8a-notification.yaml [hunk 19]
```diff
@@ -14,6 +14,8 @@ objects:
     annotations:
       description: f8a-notification-scheduler
   spec:
+    successfulJobsHistoryLimit: 1
+    failedJobsHistoryLimit: 1
```
Comment (msrb) [inline openshift/f8a-notification.yaml]: Nice :wink: Did it help?

Comment (GeetikaBatra) [reply openshift/f8a-notification.yaml]: Yep :D, but we found a bug in Openshift, looks like even though `failedJobsHistoryLimit` is set to `1`. It never deletes the older failed pods.
screenshot from 2018-08-10 11-50-29

File: openshift/helpers.sh (hunk 16 above)
Comment (GeetikaBatra) [reply openshift/helpers.sh]: I am using deploy.sh script to run the cronjob on devcluster, it is a helper script for `deploy.sh`

File: f8a_notification/entryscript.py (hunk 17 above)
Comment (GeetikaBatra) [reply f8a_notification/entryscript.py]: This is a temporary function provided to test if the script works with docker and openshift. @yzainee's changes will override this once this PR is merged. :)

File: openshift/f8a-notification.yaml [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+    schedule: "*/1 * * * *"  
+    jobTemplate:             
+      spec:
+        template:
+          spec: 
+            containers:
+            - name: f8a-notification-scheduler
+              # image: "${DOCKER_REGISTRY}/${DOCKER_IMAGE}:${IMAGE_TAG}"
+              image: registry.hub.docker.com/gb21/yusuf_notify:latest
```
Comment (msrb) [inline openshift/f8a-notification.yaml]: Do we want to keep this here?

File: openshift/f8a-notification.yaml [hunk 21]
```diff
@@ -0,0 +1,47 @@
+apiVersion: v1     
+kind: Template
+labels:
+  template: f8a-notification-scheduler
+metadata:
+  name: f8a-notification
```
Comment (msrb) [inline openshift/f8a-notification.yaml]: Please rename to `f8a-notification-scheduler`

File: openshift/f8a-notification.yaml [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+  annotations:
+    description: f8a-notification
```
Comment (msrb) [inline openshift/f8a-notification.yaml]: Please rename to `f8a-notification-scheduler`

File: .gitignore (hunk 9 above)
Comment (msrb) [reply .gitignore]: @GeetikaBatra please add also `*.pyc` to .gitignore

File: Makefile [hunk 23]
```diff
@@ -0,0 +1,28 @@
+REGISTRY?=registry.devshift.net
+REPOSITORY?=fabric8-analytics/f8a-analytics-notification-scheduler
```
Comment (msrb) [inline Makefile]: Wrong image name?

File: README.md [hunk 24]
```diff
@@ -1 +1,118 @@
...
+
+* docker >= 1.10.0
+
+Fedora 24, 25 and 26 have docker-compose > 1.6 and docker > 1.10.0. You should be
+able to run on Fedora 24/25/26 without any workarounds.
+
+Then run:
+
+```
+$ docker build . -t <username>/notify
```
Comment (msrb) [inline README.md]: Wouldn't be better to recommend `make docker-build` here?

**Comment by msrb** [comment]
@GeetikaBatra looks like you don't have git configured properly on your laptop. All commits are from `Redhat <redhat@localhost.localdomain>`.

File: README.md [hunk 25]
```diff
@@ -1 +1,118 @@
...
+
+First of all, clone the `fabric8-analytics-notification-scheduler` repo (this one).
+This includes all the configuration for running the whole system as well as some helper
+scripts and docs.
+
+### Running via docker
+
+Requirements:
+
+* docker >= 1.10.0
```
Comment (msrb) [inline README.md]: Why do we need at least this version (?)

File: openshift/deploy.sh [hunk 26]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash -e
```
Comment (msrb) [inline openshift/deploy.sh]: The script is missing executable bit.

File: requirements.in [hunk 27]
```diff
@@ -0,0 +1,5 @@
+semantic_version
+pytz
+datetime
+flask
+requests
```
Comment (msrb) [inline requirements.in]: @GeetikaBatra Please double-check these requirements. It seems to me that some of them are not really needed.

File: openshift/deploy.sh (hunk 26 above)
Comment (GeetikaBatra) [reply openshift/deploy.sh]: @msrb  I ran `chmod +x deploy.sh` to make it executable. Do I need to add anything else in the file.

File: openshift/f8a-notification.yaml (hunk 21 above)
Comment (GeetikaBatra) [reply openshift/f8a-notification.yaml]: Done.

File: openshift/f8a-notification.yaml (hunk 15 above)
Comment (GeetikaBatra) [reply openshift/f8a-notification.yaml]: Done

File: requirements.in (hunk 27 above)
Comment (GeetikaBatra) [reply requirements.in]: These will be needed for future use, once Yusuf integrates his changes.

File: README.md (hunk 25 above)
Comment (GeetikaBatra) [reply README.md]: @msrb, I think most new features are available after this version. I can remove this line if you say. :)

File: README.md (hunk 24 above)
Comment (GeetikaBatra) [reply README.md]: Yep, you are right. Done :)

File: .gitignore (hunk 9 above)
Comment (GeetikaBatra) [reply .gitignore]: Done

File: README.md (hunk 25 above)
Comment (msrb) [reply README.md]: Please, remove it :) I think we have something similar in worker repo as well, but the docs are quite outdated.

File: openshift/deploy.sh (hunk 26 above)
Comment (msrb) [reply openshift/deploy.sh]: That should do the trick, thanks ;)

File: requirements.in (hunk 27 above)
Comment (msrb) [reply requirements.in]: Isn't this supposed to be a cron job? Do we really need Flask? If yes, then I am very confused about the implementation :)

File: openshift/deploy.sh (hunk 26 above)
Comment (yzainee) [reply openshift/deploy.sh]: maybe something like this, which i see in other projects to apply the template
set -x
oc whoami
oc project
set +x

function oc_process_apply() {
  echo -e "\n Processing template - $1 ($2) \n"
  oc process -f $1 $2 | oc apply -f -
}

here=`dirname $0`
template="${here}/template.yaml"

oc_process_apply "$template"

File: f8a_notification/entryscript.py (hunk 17 above)
Comment (msrb) [reply f8a_notification/entryscript.py]: Ack, please ignore my comment then :wink:

File: requirements.in (hunk 27 above)
Comment (GeetikaBatra) [reply requirements.in]: @msrb umm yeah, you are right. Removing `flask` :)

File: README.md (hunk 25 above)
Comment (GeetikaBatra) [reply README.md]: @msrb, sure :)

**Comment by tisnik** [comment]
@GeetikaBatra just one nitpick - don't we agree to use Asciidoc over Markdown, even for README files?

File: openshift/deploy.sh (hunk 26 above)
Comment (msrb) [reply openshift/deploy.sh]: I'd also prefer such simpler script. It assumes that developer is already oc-logged on the cluster and has project selected.

But I don't mind if @GeetikaBatra wants to keep this more advanced version of the script :wink:

**Comment by yzainee** [comment]
@tisnik We will update the readme file as per your comment. For the time being, we will merge this PR as other part of the code needs to go in on top of it. 
@msrb FYI

**Pull request merged**